### PR TITLE
Add support for Hisilicon Hi3519AV100 DEMO board

### PR DESCRIPTION
**Pull request by prime-zeng**
Hi3519AV100 is a high-performance and low-power 4K
Smart IP Camera SoC designed for IP cameras, action cameras,
panoramic cameras, rear view mirrors, and UAVs. Hi3519A
V100 introduces H.265/H.264 encoding and decoding, with
performance up to 4K x 2K@60 fps and 1080p@240 fps.
For more information:
http://www.hisilicon.com/en/Products/ProductList/Surveillance

This patch has been tested using the following step,
1. Patch the uboot and Linux kernel with OP-TEE support if required
2. build step:
(1) make CROSS_COMPILE=arm-himix200-linux- PLATFORM=hisilicon
PLATFORM_FLAVOR=hi3519av100_demo  (OPTEE-OS build)

(2) make CROSS_COMPILE_HOST=arm-himix200-linux- (OPTEE_CLIENT build)

(3) cross_compile openssl and replace optee_test/host/libopenssl

(4) make CROSS_COMPILE_HOST=arm-himix200-linux-
CROSS_COMPILE_TA=arm-himix200-linux-
TA_DEV_KIT_DIR=../optee_os/out/arm-plat-hisilicon/export-ta_arm32
COMPILE_NS_USER=32  (OPTEE_TEST build)

3. mkimage  -A arm -T kernel -O tee -C none -d tee.bin uTee.optee
4. Boot setting in uboot:
nand read 0x22007fc0 0x100000 0x400000; /* load kernel */
tftp 0x30000000 uTee.optee;bootm 0x30000000;
5. after Linux startup, run daemon tee-supplicant
6. run xtest

Signed-off-by: Zeng Tao <prime.zeng@hisilicon.com>

### CODE REVIEW

File: MAINTAINERS [hunk 1]
```diff
@@ -76,6 +76,11 @@ R:	Linaro <op-tee@linaro.org>
 S:	Maintained
 F:	core/arch/arm/plat-poplar/
 
+Hisilicon hi3519Av100 family
```
Comment (jforissier) [inline MAINTAINERS]: s/hi3519Av100/Hi3519AV100/

File: README.md [hunk 2]
```diff
@@ -81,6 +81,7 @@ The **Maintained?** column shows:
 | [NXP i.MX8MMEVK Board](https://www.nxp.com/products/processors-and-microcontrollers/arm-based-processors-and-mcus/i.mx-applications-processors/i.mx-8-processors/i.mx-8m-mini-family-arm-cortex-a53-cortex-m4-audio-voice-video:i.MX8MMINI?lang=en&lang_cd=en&) |`PLATFORM=imx-imx8mmevk`| Yes | ![Actively Maintained](documentation/images/green.svg) |
 | [ARM Foundation FVP](https://developer.arm.com/products/system-design/fixed-virtual-platforms) |`PLATFORM=vexpress-fvp`| Yes | ![Actively Maintained](documentation/images/green.svg) |
 | [HiSilicon D02](http://open-estuary.org/d02-2)|`PLATFORM=d02`| No | ![Actively Maintained](documentation/images/green.svg) |
+| [HiSilicon Hi3519Av100 Demo Board](http://www.hisilicon.com/en/Products/ProductList/Surveillance)|`PLATFORM=hisilicon`| No | ![Actively Maintained](documentation/images/green.svg) |
```
Comment (jforissier) [inline README.md]: Please write the full identifier: `PLATFORM=hisilicon-hi3519av100_demo`

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: Please keep only:
```C
//  SPDX-License-Identifier: BSD-2-Clause
/*
 * Copyright (c) 2019, HiSilicon Technologies Co., Ltd.
 */
```

We don't need all the license boilerplate, because the SPDX license identifier describes the license terms unambiguously.

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+/*
+ * Entry points for the A53 aarch32 mode init.
+ * It is assume no stack is available when these routines are called.
+ * It is assume each routine is called with return address in LR
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: assumed (two lines)

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+#include <platform_config.h>
+
+.section .text
+.balign 4
+.code 32
+
+/*
+ * Cortex A53 aarch32 configuration early configuration
+ *
+ * Use scratables registers R0-R3.
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: "scratch registers"?

File: core/arch/arm/plat-hisilicon/main.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
+//  SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, HiSilicon Technologies Co., Ltd.
+ * All rights reserved.
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/main.c]: Same as above, mentions not needed.

File: core/arch/arm/plat-hisilicon/hi3519av100.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
+//  SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, HiSilicon Technologies Co., Ltd.
+ * All rights reserved.
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/hi3519av100.h]: Same as above, mentions not needed.

File: core/arch/arm/plat-hisilicon/platform_config.h [hunk 8]
```diff
@@ -0,0 +1,46 @@
+/*  SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2019, HiSilicon Technologies Co., Ltd.
+ * All rights reserved.
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/platform_config.h]: Same here, mentions not needed.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
+//  SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2019, HiSilicon Technologies Co., Ltd.
+ * All rights reserved.
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/psci.c]: Same here, mentions not needed.

File: core/arch/arm/plat-rcar/platform_config.h [hunk 10]
```diff
@@ -1,4 +1,4 @@
-/* SPDX-License-Identifier: BSD-2-Clause */
+// SPDX-License-Identifier: BSD-2-Clause
```
Comment (jforissier) [inline core/arch/arm/plat-rcar/platform_config.h]: Unrelated change (and wrong)

File: core/arch/arm/plat-hisilicon/platform_config.h [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+#ifndef PLATFORM_CONFIG_H
+#define PLATFORM_CONFIG_H
+
+#if  defined(CFG_HI3519AV100)
+#include <hi3519av100.h>
+#else
+#error "Not supported PLATFORM_FLAVOR"
+#endif
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/platform_config.h]: Just `#include <hi3519av100.h>` for now. #ifdefs can be added later if needed.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+{
+	return PSCI_VERSION_1_0;
+}
+
+void psci_system_reset(void)
+{
+	vaddr_t sysctrl = core_mmu_get_va(SYS_CTRL_BASE, MEM_AREA_IO_SEC);
+
+	if (!sysctrl) {
+		EMSG("no sysctrl mapping, hang here\n");
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/psci.c]: `\n` not needed

File: core/arch/arm/plat-hisilicon/psci.c [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+int psci_cpu_on(uint32_t core_idx, uint32_t entry,
+		uint32_t context_id)
+{
+	uint32_t val;
+	size_t pos = get_core_pos_mpidr(core_idx);
+	vaddr_t bootsram = core_mmu_get_va(BOOTSRAM_BASE, MEM_AREA_IO_SEC);
+	vaddr_t crg = core_mmu_get_va(CPU_CRG_BASE, MEM_AREA_IO_SEC);
+
+	if (!bootsram || !crg) {
+		EMSG("No bootsram or crg mapping\n");
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/psci.c]: `\n` not needed

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+ * No stack usage.
+ * LR store return address.
+ * Trap CPU in case of error.
+ */
+FUNC plat_cpu_reset_early , :
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: This is certainly quite specific to the Hi3519AV100 SoC, right? If so, the name of this file is not very good. I suggest `hi3519av100_plat_init.S`.

File: core/arch/arm/plat-hisilicon/sub.mk [hunk 15]
```diff
@@ -0,0 +1,4 @@
+global-incdirs-y += .
+srcs-y += main.c
+srcs-$(CFG_HI3519AV100) += a53_plat_init.S
```
Comment (jforissier) [inline core/arch/arm/plat-hisilicon/sub.mk]: As I said above, maybe rename the file.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	write32(0xdeadbeef, sysctrl + REG_SYSCTRL_RESET);
+}
+
+#ifdef CFG_BOOT_SECONDARY_REQUEST
+int psci_cpu_on(uint32_t core_idx, uint32_t entry,
+		uint32_t context_id)
+{
+	uint32_t val;
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/psci.c]: minor: default init to some value all local variables.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+	generic_boot_set_core_ns_entry(pos, entry, context_id);
+
+	val = virt_to_phys((void *)TEE_TEXT_VA_START);
+	write32(val, bootsram + REG_CPU_START_ADDR);
+	write32(0xe51ff004, bootsram + REG_CPU_START_COMMAND);
+
+	/* release secondary core */
+	val = read32(crg + REG_CPU_SUSSYS_RESET);
+	val &= ~(BIT32(25) | BIT32(1));
+	write32(val, crg + REG_CPU_SUSSYS_RESET);
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/psci.c]: minor: maybe use `io_clrbits32(crg + REG_CPU_SUSSYS_RESET, BIT32(25) | BIT32(1));`

Also: prefer use macro for these bits and the magic `0xe51ff004` above.
`#define RELEASE_CORE_MASK    (BIT32(25) | BIT32(1))`

File: core/arch/arm/plat-hisilicon/psci.c [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+	return PSCI_VERSION_1_0;
+}
+
+void psci_system_reset(void)
+{
+	vaddr_t sysctrl = core_mmu_get_va(SYS_CTRL_BASE, MEM_AREA_IO_SEC);
+
+	if (!sysctrl) {
+		EMSG("no sysctrl mapping, hang here\n");
+		while (1)
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/psci.c]: no `panic();` here?  (note it does not end in a endless wfi loop)

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+
+/*
+ * Entry points for the A53 aarch32 mode init.
+ * It is assume no stack is available when these routines are called.
+ * It is assume each routine is called with return address in LR
+ * and with ARM registers R0, R1, R2, R3 being scratchable.
+ */
+
+#include <arm.h>
+#include <arm32.h>
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: not nedded, included from **arm.h**

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+ * LR store return address.
+ * Trap CPU in case of error.
+ */
+FUNC plat_cpu_reset_early , :
+UNWIND(	.fnstart)
+	/*
+	 * Write the CPU Extended Control Register
+	 * Set the SMPEN bit, this Cortex-A53 core's register
+	 */
+	ldr	r0, =0x40
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: prefer define macros with nice labels one can understand what it is about and may prevents the need for an explicit inline comment.

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+	 */
+	mrs	r1, cpsr
+	cps	#CPSR_MODE_MON
+
+	read_scr r0
+	orr	r0, r0, #SCR_NS /* Set NS bit in SCR */
+	write_scr r0
+	isb
+
+	/* write HACTLR register */
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: minor: prefer start each inline comment sentence with a upper case char.

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+	mrc	p15, 4, r2, c1, c0, 1
+	orr	r2, r2, #1<<1
+	mcr	p15, 4, r2, c1, c0, 1
+
+	bic	r0, r0, #SCR_NS /* Clr NS bit in SCR */
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: minor: maybe remove the comment. the macro label says about the same.

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+	cps	#CPSR_MODE_MON
+
+	read_scr r0
+	orr	r0, r0, #SCR_NS /* Set NS bit in SCR */
+	write_scr r0
+	isb
+
+	/* write HACTLR register */
+	mrc	p15, 4, r2, c1, c0, 1
+	orr	r2, r2, #1<<1
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: prefer a expended form with spaces, or better use `BIT(1)`, or even better, define a macro
```
#define HACTLR_CPUECTLR_ACCESS_NSEC	BIT(1)
```

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+	 * enable cci for secondary core
+	 */
+	mov	r3, lr
+	bl	__get_core_pos
+	mov	lr, r3
+	cmp	r0, #0
+	beq	out
+	ldr	r0, =0x04528000
+	ldr	r1, [r0]
+	orr	r1, r1, #0x200       /* bit 9 set to 1 */
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: prefer `BIT(9)`

File: core/arch/arm/plat-hisilicon/a53_plat_init.S [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+	isb
+
+	/* write ACTLR register */
+	mrc	p15, 0, r2, c1, c0, 1
+	orr	r2, r2, #1<<1
+	mcr	p15, 0, r2, c1, c0, 1
+
+	msr	cpsr, r1
+	/*
+	 * enable cci for secondary core
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/a53_plat_init.S]: minor: prefer start inline comment sentences with an upper case char.

File: core/arch/arm/plat-hisilicon/hi3519av100.h [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+#define BOOTSRAM_BASE			(0x04200000)
+#define BOOTSRAM_SIZE			(0x1000)
+
+/* CPU Reset Control */
+#define CPU_CRG_BASE			(0x04510000)
+#define CPU_CRG_SIZE			(0x1000)
+
+/* Sysctrl Register */
+#define SYS_CTRL_BASE			(0x04520000)
+#define SYS_CTRL_SIZE			(0x1000)
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/hi3519av100.h]: minor: maybe remove the parentheses

File: core/arch/arm/plat-hisilicon/hi3519av100.h [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+ *  0x3260_0000 [TZDRAM_BASE, TZSRAM_BASE, TEE_LOAD_ADDR]
+ *    OP-TEE Future Use: 2 MiB
+ *  0x3240_0000
+ *    Shared memory: 4 MB
+ *  0x3200_0000
+ *    Linux memory: 256MB
+ *  0x2200_0000
+ *    DSP reserved memory:      32MB
+ *  0x2000_0000 [DRAM_BASE]
+ */
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/hi3519av100.h]: this is related to info in **conf.mk**. Maybe move this comment there.

File: MAINTAINERS (hunk 1 above)
Comment (prime-zeng) [reply MAINTAINERS]: yes, i will also update the same spelling error in README.MD

File: core/arch/arm/plat-hisilicon/a53_plat_init.S (hunk 3 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/a53_plat_init.S]: I am not sure, but the only thing i want to make sure is that we have no legal issues by doing so, would you double check it for me?   Now i only see the Linaro is doing like what you suggested, but no other SOC vendors.

File: core/arch/arm/plat-hisilicon/psci.c (hunk 18 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/psci.c]: Yes, Panic should be  the correct behavior here.

File: core/arch/arm/plat-hisilicon/psci.c (hunk 17 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/psci.c]: Good suggestion.

File: core/arch/arm/plat-hisilicon/psci.c (hunk 16 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/psci.c]: But i think it is useless due to the compiler optimization. right?

File: core/arch/arm/plat-hisilicon/a53_plat_init.S (hunk 14 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/a53_plat_init.S]: Yes, you are right, good suggestion.

File: core/arch/arm/plat-hisilicon/a53_plat_init.S (hunk 3 above)
Comment (jforissier) [reply core/arch/arm/plat-hisilicon/a53_plat_init.S]: There are commits in optee_os that use the SPDX identifier and no boilerplate license text. For instance, this one by NXP: https://github.com/OP-TEE/optee_os/commit/ad22421.

The Linux kernel does the same. See for instance:
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=09d290f0ba21
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=2849dd8bc72b6

This is explained at: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/license-rules.rst?h=v5.0-rc7#n40, more precisely: "An alternative to boilerplate text is the use of [SPDX] identifiers".

So I really think it is OK to not add the license text.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+
+	if ((pos == 0) || (pos >= CFG_TEE_CORE_NB_CORE))
+		return PSCI_RET_INVALID_PARAMETERS;
+
+	/* set secondary core' NS entry addresses */
+	generic_boot_set_core_ns_entry(pos, entry, context_id);
+
+	val = virt_to_phys((void *)TEE_TEXT_VA_START);
+	write32(val, bootsram + REG_CPU_START_ADDR);
+	write32(0xe51ff004, bootsram + REG_CPU_START_COMMAND);
```
[the remark this answers is not in the document]
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/psci.c]: ok, will fix in a patch.

File: core/arch/arm/plat-hisilicon/hi3519av100_plat_init.S [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+	msr	cpsr, r1
+	/*
+	 * Enable cci for secondary core
+	 */
+	mov	r3, lr
+	bl	__get_core_pos
+	mov	lr, r3
+	cmp	r0, #0
+	beq	out
+	ldr	r0, =0x04528000
```
[the remark this answers is not in the document]
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/hi3519av100_plat_init.S]: ok, will fix in a new patch.

File: core/arch/arm/plat-hisilicon/hi3519av100_plat_init.S [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+	cps	#CPSR_MODE_MON
+
+	read_scr r0
+	orr	r0, r0, #SCR_NS /* Set NS bit in SCR */
+	write_scr r0
+	isb
+
+	/* Write HACTLR register */
+	mrc	p15, 4, r2, c1, c0, 1
+	orr	r2, r2, #1<<1
```
[the remark this answers is not in the document]
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/hi3519av100_plat_init.S]: ok, will fix in a new patch.

File: core/arch/arm/plat-hisilicon/psci.c (hunk 18 above)
Comment (jforissier) [reply core/arch/arm/plat-hisilicon/psci.c]:
```C
if (!sysctrl)
	panic("No sysctrl mapping");
```

**Comment by prime-zeng** [comment]
All the review foundings are fixup now, but i can't do the rebase and upload to the github because of the proxy rules,  please help to check if there is any pending issues, if everything 's ok, it will squash the commits,  thanks.   @jforissier  @etienne-lms

**Comment by prime-zeng** [comment]
@jforissier @etienne-lms  Thanks very much for your views,  i think all the founding are targeted now, please help to confirm if there are something missing?  If everything is ok, i will squash the commits, by the way,  I can't do the rebase and push it to github because of the proxy limit, does it matter or not?

**Comment by jforissier** [comment]
@prime-zeng no need to rebase, it is done automatically when I merge the PR anyway.
However, if I understand well, you are using `io_write32()` which has been introduced in a later commit that you don't have in your repository (since you said you could not rebase)? So, how have you tested the patches?

Before squashing the commits I suppose you can wait for @etienne-lms R-B or A-B tag.

**Comment by prime-zeng** [comment]
I create a new branch and rebase locally, and then do the test.

**Comment by jforissier** [comment]
> I create a new branch and rebase locally, and then do the test.

OK, if this is tested then all is well :) Thanks.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+
+	if (!bootsram || !crg) {
+		EMSG("No bootsram or crg mapping");
+		return PSCI_RET_INVALID_PARAMETERS;
+	}
+
+	if ((pos == 0) || (pos >= CFG_TEE_CORE_NB_CORE))
+		return PSCI_RET_INVALID_PARAMETERS;
+
+	/* set secondary core' NS entry addresses */
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/psci.c]: minor: maybe remove character `'`.

File: core/arch/arm/plat-hisilicon/psci.c [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+#include <tee/entry_std.h>
+#include <tee/entry_fast.h>
+
+#define REG_CPU_SUSSYS_RESET	0xcc
+#define REG_CPU_START_COMMAND	0x0
+#define REG_CPU_START_ADDR	0x4
+#define REG_SYSCTRL_RESET	0x4
+#define RELEASE_CORE_MASK	(BIT32(25) | BIT32(1))
+
+
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/psci.c]: minor: maybe remove one of the empty line.

File: core/arch/arm/plat-hisilicon/main.c [hunk 33]
```diff
@@ -0,0 +1,64 @@
...
+#include <kernel/pm_stubs.h>
+#include <mm/tee_pager.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <tee/entry_std.h>
+#include <tee/entry_fast.h>
+
+#ifdef GIC_BASE
+static struct gic_data gic_data;
```
Comment (etienne-lms) [inline core/arch/arm/plat-hisilicon/main.c]: `gic_data` is not used and `GIC_BASE`  is not defined. Can remove this definition.

File: core/arch/arm/plat-hisilicon/psci.c (hunk 31 above)
Comment (prime-zeng) [reply core/arch/arm/plat-hisilicon/psci.c]: Maybe `core's` is better

**Comment by prime-zeng** [comment]
Since the README have been moved to optee_docs,  create a PR to target it, please refer to https://github.com/OP-TEE/optee_docs/pull/1
